Transform JS values to Lisp ones.

aartaka opened the pull request:
This includes a `jsc-value-to-lisp` function to transform `JSCValue`s to lispy versions of these, and necessary changes to the codebase to handle these.

# What's There:
- `jsc-value-to-lisp`.
- Special variables (see Things to Discuss regarding these): 
  - `*js-null-value*` for what to translate `null` to.
  - `*js-undefined-value*` for what to translate `undefined` to.
  - `*js-true-value*` for what to translate `true` to.
  - `*js-false-value*` for `false`.
  - `*js-function-value*` for functions.
    - This exists because WebKit/JSCore functions managinf function objects are not yet bound, thus functions are balck boxes to us.
  - `*js-array-type*` for what to translate arrays to -- lists or vectors. 
  - `*js-object-type*` for what to translate objects to -- alists, plists, or hash-tables. 
    - I considered CLOS objects at some moment, but that seems to be an overkill.
- `JSCExceptions` are handled properly, although I cannot test it right because of `GError`s popping out on any erroneous JS instead of `JSCExceptions`.

# Things to Discuss
- I want special variables to be `let`-rebindable, i.e. `(let ((webkit:*js-object-type* :alist)) (jsc-value-to-lisp ...))` shoud turn any object into an alist instead of default hash-table. I don't yet understand how to do that (I miss dynamic extent from Elisp from time to time, yes).
- Should function bindings be in this patch? I'm pretty dure we can live without them, but there could be reasons I've not considered.
- After this patch, we need to bump version up, because the behavior of `javascript-evaluation-complete` is changed dramatically.
  - Nyxt will need to adjust too. I can do necessary changes to Nyxt once we merge this. The transition won't be easy, though.
- We still cannot move e.g., `document` to Lisp land (as discussed in #56) -- WebKit errors obscurely. Our arsenal is much wider now, though: `null`, `undefined`, booleans, numbers, strings, arrays, classless objects, and any combinations of those. Our own Lispy JSON!
  - Numbers are always translated to floats, even if those are exact. Should we be smart about that?
  - Are default types fo arrays (list) and objects (hash-table) sane? It feels like most Lisp-native structures that cannot be mistaken for one another.

# What to Do Next:
- Add JS function binding and translate those to Lisp somehow (lambdas wrapping the `jsc_value_function_call`?)
- Make necessary changes to Nyxt and use the full potential of this patch (e.g., in Plump DOM).
- Add class management (this can allow us to pull `document` in, for example. No guarantees, though.)

How does it look to you?

Ambrevar commented on webkit2/jscore.jsc-exception.lisp:
```diff
@@ -64,3 +64,13 @@
 (export 'jsc-exception-report)
+
+(define-condition jsc-exception-condition (error)
+  ((exception :initform (error "There should be a JSCException.")
```
The error is a bit confusing, because we could believe that it's _about_ a JS error :)
Maybe "Required argument..." instead.

aartaka replied:
Fixed in 9d7bdf78a785a26a9b26d0e9f0cd93c16fcd84d9.

Ambrevar commented on webkit2/jscore.jsc-exception.lisp:
```diff
@@ -64,3 +64,13 @@
+              :initarg :exception))
+  (:documentation "Condition to signal when a JSCException is raised.")
+  (:report (lambda (condition stream)
+             (format stream "~a~&~%~%~a"
```
Why not just "~&~&"?

aartaka replied:
Hmmmm...

aartaka replied:
Simplified in e213c8ff3fbe962cf0d893644e3ce71509760547.

Ambrevar commented on webkit2/jscore.jsc-value.lisp:
```diff
@@ -164,3 +164,115 @@
+
+(defvar *js-false-value* nil
+  "The value used when translating false from JavaScript to Lisp.")
+(export '*js-false-value*)
```
The choice between nil for `false` and nil for `null` is delicate.

Any idea what other CL JSON libraries do?

aartaka replied:
The [blogpost you mentioned in #56](https://sabracrolleton.github.io/json-review) lists `nil` as a more widespread value for falsity (5 out of 7 libraries do that), than for nullness (3 or 4 out of 7). Given that using `nil` for both is not an option, it'd better be a boolean value, along with t for truth.

Ambrevar commented on webkit2/jscore.jsc-value.lisp:
```diff
@@ -164,3 +164,115 @@
+- false to FALSE-VALUE (nil by default). Also see `*js-false-value*'.
+- null to NULL-VALUE (:null by default). Also see `*js-null-value*'.
+- undefined to UNDEFINED-VALUE (:undefined by default). Also see `*js-undefined-value*'.
+- JS functions to FUNCTION-VALUE (:function by default). This will change in the future!
```
Why will it change?

aartaka replied:
Because I want to leverage WebKit to actually call JS from Lisp.

Ambrevar replied:
So same as the export above, if we are unsure about the API, then I suggest we don't document the return value of the function.
Or maybe document at outside of this list and highlight that's it's an unsupported return value.

Ambrevar commented:
> I don't yet understand how to do that.

Looks like what you did is correct, what's wrong?

Ambrevar commented on webkit2/jscore.jsc-value.lisp:
```diff
@@ -164,3 +164,115 @@
+                               (cons (cffi:foreign-string-to-lisp property-name) property-value)))
+                   into properties
+                 finally (return
+                           (prog1
```
Could just be a `progn` with `properties last, no?

aartaka replied:
Fair. I've just liked `prog1` more there :)

I've changed it to `progn` in 8b8e221b90a2fc57916d96d4954a4e3e03253c18.

Ambrevar commented:
> Should function bindings be in this patch? I'm pretty dure we can live without them, but there could be reasons I've not considered.

You mean JS function -> CL conversion?  If so, how would we handle JS functions?

Ambrevar commented:
> After this patch, we need to bump version up, because the behavior of javascript-evaluation-complete is changed dramatically.

About time!  The version hasn't been bumped in ages.

> Nyxt will need to adjust too. I can do necessary changes to Nyxt once we merge this. The transition won't be easy, though.

I think it's OK to do between 2 stable versions.

Ambrevar commented:
> Numbers are always translated to floats, even if those are exact. Should we be smart about that?

As far as I know, JS only has the "number" type.  That said, we could make it automatic number conversion an option just like we can choose :list or :vectors.

aartaka commented:
> > I don't yet understand how to do that.
> 
> Looks like what you did is correct, what's wrong?

It just doesn't work. For example, running this piece of code:
``` lisp
(progn
  (defvar *js-result*)
  (let ((webkit:*js-object-type* :plist))
   (nyxt::%within-renderer-thread-async
    (lambda ()
      (webkit2:webkit-web-view-evaluate-javascript
       (nyxt::gtk-object (nyxt:current-buffer))
       "var obj = {whetever: 3}; obj"
       (lambda (object) (setf *js-result* object)) #'nyxt::javascript-error-handler))))
  *js-result*)
```
returns a hash-table, instead of plist.

Ambrevar commented:
> Add JS function binding and translate those to Lisp somehow (lambdas wrapping the jsc_value_function_call?)

Then you'd need to convert CL arguments to JS, no?  Would be cool though, but also quite some work.
Do you know an example of where it's needed?

aartaka commented:
> > Numbers are always translated to floats, even if those are exact. Should we be smart about that?
> 
> As far as I know, JS only has the "number" type. That said, we could make it automatic number conversion an option just like we can choose :list or :vectors.

You're right, JS has only one numeric type, and it's IEEE 754 float. Optional rounding is an option, though.

Ambrevar commented:
>   (nyxt::%within-renderer-thread-async

I think your issue is with threads.  Threads reset the dynamic bindings, you must perform the binding either inside the thread or in the key argument of `make-thread`.

aartaka commented:
> > Add JS function binding and translate those to Lisp somehow (lambdas wrapping the jsc_value_function_call?)
> 
> Then you'd need to convert CL arguments to JS, no? Would be cool though, but also quite some work.

No need to convert to JS directly. Once we convert it to C/Glib/WebKitGTK values, WebKit will do the rest.

> Do you know an example of where it's needed?

WebExtensions _need_ this.

aartaka commented:
> > (nyxt::%within-renderer-thread-async
> 
> I think your issue is with threads. Threads reset the dynamic bindings, you must perform the binding either inside the thread or in the key argument of `make-thread`.

This
``` lisp
(progn
  (defvar *js-result*)
  (nyxt::%within-renderer-thread-async
   (lambda ()
     (let ((webkit:*js-object-type* :plist))
       (webkit2:webkit-web-view-evaluate-javascript
        (nyxt::gtk-object (nyxt:current-buffer))
        "var obj = {whetever: 3}; obj"
        (lambda (object) (setf *js-result* object)) #'nyxt::javascript-error-handler))))
  *js-result*)
```

doesn't work either.

aartaka commented:
> > Should function bindings be in this patch? I'm pretty dure we can live without them, but there could be reasons I've not considered.
> 
> You mean JS function -> CL conversion? If so, how would we handle JS functions?

Making CL functions calling necessary JS function via WebKit. E.g.
``` lisp
(cond 
...
((jsc-value-is-function jsc-value) (lambda (&rest args) (jsc-value-function-call jsc-value (transform-to-webkit-values args)))))
...)
```

Ambrevar commented:
> ``` lisp
> (progn
>   (defvar *js-result*)
>   (nyxt::%within-renderer-thread-async
>    (lambda ()
>      (let ((webkit:*js-object-type* :plist))
>        (webkit2:webkit-web-view-evaluate-javascript
>         (nyxt::gtk-object (nyxt:current-buffer))
>         "var obj = {whetever: 3}; obj"
>         (lambda (object) (setf *js-result* object)) #'nyxt::javascript-error-handler))))
>   *js-result*)
> ```
>
> doesn't work either.

Hmmm, isn't it because jsc-value-to-lisp is call here:

> (funcall (callback-function callback) (jsc-value-to-lisp value)))

Which is run in a separate thread?  Maybe the dynamic binding is lost at
some point.  Note that the above line is run in
javascript-evaluation-complete which is a callback to
webkit-web-view-evaluate-javascript.

Have you tried putting some stickers, or maybe break and inspect the
dynamic value in different parts of `webkit-web-view-evaluate-javascript`.

aartaka commented:
> ``` lisp (progn (defvar *js-result*) (nyxt::%within-renderer-thread-async (lambda () (let ((webkit:*js-object-type* :plist)) (webkit2:webkit-web-view-evaluate-javascript (nyxt::gtk-object (nyxt:current-buffer)) "var obj = {whetever: 3}; obj" (lambda (object) (setf *js-result* object)) #'nyxt::javascript-error-handler)))) *js-result*) ``` doesn't work either.
> Hmmm, isn't it because jsc-value-to-lisp is call here:
> (funcall (callback-function callback) (jsc-value-to-lisp value)))
> Which is run in a separate thread? Maybe the dynamic binding is lost at some point. Note that the above line is run in javascript-evaluation-complete which is a callback to webkit-web-view-evaluate-javascript. Have you tried putting some stickers, or maybe break and inspect the dynamic value in different parts of `webkit-web-view-evaluate-javascript`.

Indeed, it is called in a yet another thread, so let-binding is no help there. I guess that's something to live with -- we just need to remember `setf`-ing necessary variables then :)

aartaka commented:
> > > Should function bindings be in this patch? I'm pretty dure we can live without them, but there could be reasons I've not considered.
> > 
> > 
> > You mean JS function -> CL conversion? If so, how would we handle JS functions?
> 
> Making CL functions calling necessary JS function via WebKit. E.g.
> 
> ```lisp
> (cond 
> ...
> ((jsc-value-is-function jsc-value) (lambda (&rest args) (jsc-value-function-call jsc-value (transform-to-webkit-values args)))))
> ...)
> ```

Beware that this example is purely fictional -- `jsc-value-function-call` is not yet bound, has a different argument list, and does not accept Lisp values. `transform-to-webkit-values` does not exist either.

aartaka commented:
All addressed!

jmercouris commented:
I just wanted to briefly chime in and say WOW! WOW! :-)

Looking forward to seeing this merged.

Ambrevar commented on webkit2/jscore.jsc-value.lisp:
```diff
@@ -164,3 +164,115 @@
+(defvar *js-function-value* :function
+  "The value used when translating function values from JavaScript to Lisp.
+Likely to get deprecated.")
+(export '*js-function-value*)
```
If we are not sure about the API, let's not export it then?

Ambrevar commented:
Before merging this, I believe this changeset begs for tests!
They will be useful to guarantee the integrity of the conversion over time.

@aartaka: If you feel lazy to write them, let me know, I can give a hand.

aartaka commented:
OK, function variable is unexported.

aartaka commented:
Regarding tests: I've written yet another test browser function, but I cannot make it to work properly (it doesn't append all the necessary HTML). Here's the code:
``` lisp
(defun js-transform-browser-main ()
  "A JS transformation-checking version of `simple-browser-main'."
  (gtk:within-main-loop
    (let* ((win (make-instance 'gtk:gtk-window))
           (context (make-instance 'webkit:webkit-web-context))
           (view (make-instance 'webkit2:webkit-web-view
                                :web-context context)))
      (macrolet ((with-js-transform-tests (view &body clauses)
                   `(progn
                      ,@(loop for (js-string lisp-value) in clauses
                              collect
                              `(webkit2:webkit-web-view-evaluate-javascript
                                ,view
                                ,js-string
                                (lambda (object)
                                  (webkit2:webkit-web-view-evaluate-javascript
                                   ,view
                                   (let ((success (equalp ,lisp-value object)))
                                     (format nil "document.querySelector(\"ul\").insertAdjacentHTML(\"beforeend\", \"<li style=\"color=~:[red~;green~]\">~:[X~;✓~] ~s = ~a</li>\");"
                                             success success
                                             ,js-string ,lisp-value)))))))))
        (gobject:g-signal-connect win "destroy"
                                  #'(lambda (widget)
                                      (declare (ignore widget))
                                      (gtk:leave-gtk-main)))
        (gtk:gtk-container-add win view)
        (webkit2:webkit-web-view-load-uri view "about:blank")
        (gtk:gtk-widget-show-all win)
        (webkit2:webkit-web-view-evaluate-javascript
         view
         "document.body.innerHTML = \"<h1>Testing Javascript values transformation</h1><ul></ul>\""
         (lambda (object)
           (declare (ignore object))
           (with-js-transform-tests view
             ("5" 5)
             ("\"hello\"" "hello")
             ("[1, 2, 3, 4, 5]" '(1 2 3 4 5)))))))))
```

I'm not lazy about tests, I'm just not sure how to approach this deeply intertwined interoperation, except the non-working snippet above.

aartaka commented:
Testing it on Nyxt side is more or less easy, though.

aartaka commented:
Testing: done. See 156687dc1a38e568dff7e1768e44c50c05751ba1. It wasn't trivial at all.

aartaka commented:
Testing have revealed that transformation of functions throws errors much like `document` transformation, but I guess that's OK.

aartaka commented:
There seems to be a range of things WebKit just cannot allow transforming ¯\_(ツ)_/¯

aartaka commented:
OK, functions are only recognized as such if they were created by the programmer using `webkit_jsc_value_new_function()`. Otherwise it errors out. We have to live with it.

I've removed the function mentions from the `js-value-to-lisp` docstring and moved its `cond` clause to the very bottom. Is that enough of instability mark? :)

Should be alright to merge otherwise.

Ambrevar commented:
OK for the exports.

You were saying this is not backward compatible.  Why?
If this is really the case, then we should bump the version.  I suggest 3.0.0 then.  Thoughts?

About the test: good start, I would suggest an actual test suite though (so that we can call `asdf:test-system` on `cl-webkit`).  And test all conversions one by one, with different rules (hash-table / plist / alist for objects, etc.).
Sorry if this is a bit of a pain, but considering how important this is for the future of cl-webkit, I believe we ought to have a robust foundation :)

aartaka commented:
> You were saying this is not backward compatible. Why?

It's not backwards-compatible because we pass arbitrary Lisp values into JS evaluation callbacks, instead of string. Testing on Nyxt, it breaks almost everything JS-reliant.

> If this is really the case, then we should bump the version. I suggest 3.0.0 then. Thoughts?

Sounds alright. I though about adding function bindings before we bump the version, though. Maybe it'll be better to bind function-related methods as a part of this patch. I'm not sure. What do you think -- maybe we can merge this and bind functions afterwards, and bump version once we have function bindings?

> About the test: good start, I would suggest an actual test suite though (so that we can call `asdf:test-system` on `cl-webkit`). And test all conversions one by one, with different rules (hash-table / plist / alist for objects, etc.).

Hmmmm. This requires setting an environment (windows, views, contexts) in a test suite. Is there a testing library that'd allow that? Will good old `cl-prove` work there?

> Sorry if this is a bit of a pain, but considering how important this is for the future of cl-webkit, I believe we ought to have a robust foundation.

It's totally fine, I understand that. `cl-webkit` was living without tests for too long :) I'll dig `cl-prove` and write proper tests.

Ambrevar commented:
> Sounds alright. I though about adding function bindings before we bump the version, though. Maybe it'll be better to bind function-related methods as a part of this patch. I'm not sure. What do you think -- maybe we can merge this and bind functions afterwards, and bump version once we have function bindings?

Not sure we can, because cl-webkit hasn't had a version bump in a very
long time, which means everyone is following master by now (including
Quicklisp).

Thus the version bump should happen at the same time as backward
compatibility breaks.

>> About the test: good start, I would suggest an actual test suite though (so that we can call `asdf:test-system` on `cl-webkit`). And test all conversions one by one, with different rules (hash-table / plist / alist for objects, etc.).
>
> Hmmmm. This requires setting an environment (windows, views, contexts) in a test suite. Is there a testing library that'd allow that? Will good old `cl-prove` work there?

You can define some setup code as part of your test suite.  Is there
something specific you'd be expecting from your test suite?

Pick your favourite library :)

>> Sorry if this is a bit of a pain, but considering how important this is for the future of cl-webkit, I believe we ought to have a robust foundation.
>
> It's totally fine, I understand that. `cl-webkit` was living without tests for too long :) I'll dig `cl-prove` and write proper tests.

If you want to give a go to fiveam, parachute or anything else, go ahead :)

aartaka commented:
OK, I tried doing tests with `fiveam` and it's going alright, except for one thing -- multithreading. Having no Calispel/Chanl/lparralel or any other way to do a blocking thread communication, I'm forced to re-invent the wheel. No success so far, although I'm getting close.

I've pushed the draft of the testing code to the branch, you can review and point at what I missed in this multithreading spaghetti :) Beware: `fiveam` opens debugger on failing tests, and basically all of the tests in this patch are failing because of broken shared state :D 

In any way, I'll get back to it tomorrow, so that my head is fresh. Binding JS function creation and calling in the meanwhile.

Ambrevar commented:
I believe it's OK to used Calispel in the cl-webkit/test system :)

aartaka commented:
> I believe it's OK to used Calispel in the cl-webkit/test system :)

Alright, I'll push tests with Calispel in five hours then.

jmercouris commented:
I want to expand upon what Pierre said- Adding dependencies for tests can be done with a lot less friction than adding dependencies for the main system.

aartaka commented:
Tests added, several plist- and float-related bugs fixed. Wait for the function bindings and we should be good to merge!

Ambrevar commented on webkit2/cl-webkit2.asd:
```diff
@@ -76,4 +76,14 @@
+
+(defsystem :cl-webkit2/test
+  :description "CL-WebKit tests, mainly to check JS result conversion."
+  :license "MIT"
```
Probably not necessary to repeat the license.

Ambrevar commented on webkit2/cl-webkit2.asd:
```diff
@@ -76,4 +76,14 @@
+  :pathname "tests/"
+  :components ((:file "package")
+               (:file "tests"))
+  :perform (test-op (o c) (symbol-call :5am :run! :webkit2.test)))
```
Today I learnt about `symbol-call` :)

Ambrevar commented on webkit2/tests/tests.lisp:
```diff
@@ -0,0 +1,258 @@
+     (let ((,var (calispel:? channel)))
+       ,@body)))
+
+(gtk:within-main-loop
```
This should belong to a setup function, no?  Otherwise it would start the GTK loop when asdf:load-ing.

Ambrevar commented on webkit2/tests/tests.lisp:
```diff
@@ -0,0 +1,258 @@
+  (with-js-transform-result (%result% "-Math.pow(10, 100)")
+    (is (= -1.0000000000000002d100 %result%))))
+
+;; REVIEW: CCL? ECL?
```
Found it! https://shinmera.github.io/float-features/

Ambrevar commented on webkit2/tests/tests.lisp:
```diff
@@ -0,0 +1,258 @@
+
+(def-test periodic-number (:suite js-tests)
+  (with-js-transform-result (%result% "var num = 5/3; num")
+    (is (equal 1.6666666666666667d0 %result%))))
```
Is the floating point representation portable?

Ambrevar commented on webkit2/tests/tests.lisp:
```diff
@@ -0,0 +1,258 @@
+  (setf webkit:*js-array-type* :vector)
+  (with-js-transform-result (%result% "[true, false, undefined, null, 100000, \"hello\", {one: 1}]")
+    (is (equalp (vector t nil :undefined :null 100000 "hello" '(("one" . 1))) %result%)))
+  (setf webkit:*js-array-type* :list))
```
Use unwind-protect to ensure the global is reset on error.

Ideally, we should figure out the dynamic binding properly.

Ambrevar commented on webkit2/tests/tests.lisp:
```diff
@@ -0,0 +1,258 @@
+                 ("nul" . :null)
+                 ("undef" . :undefined)
+                 ("googol" . 1.0000000000000002d100)
+                 ("nil") ;; Maybe use non-dotted alists instead of this?
```
Why?  Can't you put `"nil" . nil` explicitly to emphasize it's `false`?

Ambrevar commented on webkit2/tests/tests.lisp:
```diff
@@ -0,0 +1,258 @@
+
+(def-suite js-tests :description "Testing JS value transformation." :in webkit-tests)
+
+(defvar *webkit-environment* (make-hash-table :test 'equal))
```
Why a hash table?  Don't you want to make a class/struct instead?

aartaka commented:
What's new:
- JS function definitions and tests for those (`make-jsc-function` macro).
- `lisp-to-jsc-value` function that inverts the work of `jsc-value-to-lisp` perfectly (there's even a test checking the equality of results!)
- `float-features` used whereever necessary.
- Numerous fixes.
- `webkit-web-view-evaluate-javascript` now accepts an optional argument denoting the world to evaluate the JavaScript code in.

Bad thing is, neither CCL nor ECL cannot finish running test suite somewhy, and SBCL throws package variance errors on my machine. 

If we can fix tests to work at least in CCL, then this patch should be mergeable (I planned to add JS classes interop, but those can wait).

aartaka commented:
How about merging this without CCL support for tests (it may well be working for usual purposes beyond quirky multithreaded mess of the tests)? I'll start working on Nyxt side of things right after we merge then.

Ambrevar commented:
Does CCL work for the demo?

What exactly is the package variance issue with SBCL?  Do you have a recipe?

We also need to bump the version to 3.0.0 I think, and add a change log.

aartaka commented:
> Does CCL work for the demo?

Yes, it does. We can merge it then, I guess.
 
> What exactly is the package variance issue with SBCL? Do you have a recipe?

Forget it, it's some issue with my environment.

> We also need to bump the version to 3.0.0 I think, and add a change log.

Sure thing! See cb35320d45d68e86c36d0e712479650e58e391fa and b282f165fba2af9b0b65044929328bb98a1e8306. That enough?

Ambrevar commented:
Looking good, let me test one last time!

After this, I'll update the Guix package.

aartaka commented:
Classes are there! Let me bind methods (won't take much time, they are essentially the same as function), and we can merge.

aartaka commented:
Methods and collateral fixes are there too. It's complete, anything I can add on top is occasional bugfixes and fancy backwards-compatible features.

Ambrevar commented on webkit2/jsc-utils.lisp:
```diff
@@ -0,0 +1,397 @@
+           jsc-value
+           (length args)
+           (if args
+               ;; TODO: Free this.
```
If you want to fix this later, open an issue about it.

Ambrevar commented on webkit2/cl-webkit2.asd:
```diff
@@ -20,7 +20,7 @@
   :author "Joachim Fasting <joachifm@fastmail.fm>"
   :licence "MIT"
-  :version "0.0"
+  :version "3.0"
```
What about using semver from now on and thus 3.0.0 instead?

aartaka replied:
Done in 4a72f251c17566a82d494f11af1f4cea79c58c3f.

Ambrevar commented on demo/simple-browser.lisp:
```diff
@@ -121,11 +121,70 @@ Loads and renders a single web page."
+                              `(webkit2:webkit-web-view-evaluate-javascript
+                                ,view
+                                ,js-string
+                                (lambda (object jsc-object)
```
Compiling this file raises a warning:

```
The variable JSC-OBJECT is defined but never used.
```

Ambrevar commented on demo/simple-browser.lisp:
```diff
@@ -121,11 +121,70 @@ Loads and renders a single web page."
 
-(defun test-browser-main (&key private extended styled custom-scheme)
+(defun js-transform-browser-main ()
+  "A JS transformation-checking version of `simple-browser-main'."
```
Maybe "validating" instead of "checking" and "variation" / "variant" instead of "version".

Ambrevar commented on README.org:
```diff
@@ -45,6 +51,17 @@ rights to the work you contribute.
+  results of code evaluation, and an original [[https://webkitgtk.org/reference/jsc-glib/2.32.1/JSCValue.html][JSCValue]] it was
+  transformed from.
+- Add tests (webkit2/tests/ subdirectory and =cl-webkit2/tests= system),
+  mostly for JS evaluation results.
```
Nit: a newline would be nice here.

Ambrevar commented on README.org:
```diff
@@ -45,6 +51,17 @@ rights to the work you contribute.
+- The callback to the =webkit-web-view-evaluate-javascript= now accepts
+  two required arguments -- the Lispy data, transformed from the
+  results of code evaluation, and an original [[https://webkitgtk.org/reference/jsc-glib/2.32.1/JSCValue.html][JSCValue]] it was
+  transformed from.
```
This got me to think, what if the callback took `(jsc-value & optional lisp-value)` instead?  Wouldn't be backward compatible?

aartaka replied:
Won't work. Both `(funcall (lambda (a) a) 1 2)` and `(apply (lambda (a) a) 1 2 nil)` signal errors. We use strict one-argument callbacks like `(lambda (a) a)` everywhere, thus this code will become erroneous even with optional arguments. Thus, it's inevitably backwards incompatible.

Ambrevar commented on webkit2/jsc-utils.lisp:
```diff
@@ -0,0 +1,397 @@
+    (declare (ignore value))
+    (jsc-value-new-undefined context))
+  (:documentation "Transform a provided Lisp value to the most sensible JSCValue counterpart.
+In case no suitable method was found, create a JSCValue for undefined."))
```
Put `undefined` between quotes, otherwise it looks like broken English :)

Ambrevar commented on webkit2/jsc-utils.lisp:
```diff
@@ -0,0 +1,397 @@
+
+(export 'make-jsc-function)
+(defmacro make-jsc-function ((view &optional name context-designator) args &body body)
+  "Create a new JSCValue function with a callback having ARGS and BODY.
```
Which callback?
"having" -> "over"

Ambrevar commented on webkit2/jsc-utils.lisp:
```diff
@@ -0,0 +1,397 @@
+- String names new function directly.
+- nil gives no name to the new function, making it anonymous.
+
+See `get-jsc-context' for what CONTEXT-DESIGNATOR could be."
```
Maybe "...for CONTEXT-DESIGNATOR allowed values".  Same below.

Ambrevar commented on webkit2/jsc-utils.lisp:
```diff
@@ -0,0 +1,397 @@
+
+NAME is either a symbol, a string, or nil:
+- Symbol turn into a camelcase JS function name.
+- String names new function directly.
```
"A symbol is turned..."
"A string names the new function directly."

Ambrevar commented on webkit2/jsc-utils.lisp:
```diff
@@ -0,0 +1,397 @@
+PARENT-CLASS should be another JSCClass.
+SLOTS are a (possibly empty) list of entries of a form:
+\(SLOT-NAME &KEY READER WRITER), where
+- SLOT-NAME could be a symbol or a string. Either way, it's passed to
```
"could" -> "can"

Ambrevar commented on webkit2/jsc-utils.lisp:
```diff
@@ -0,0 +1,397 @@
+- SLOT-NAME could be a symbol or a string. Either way, it's passed to
+  JSCClass as string. In case it's a symbol, it's transformed to a
+  camelcase property name.
+- READER should be funcall-able, so it should either be:
```
"should" -> "must"

Ambrevar commented on webkit2/jsc-utils.lisp:
```diff
@@ -0,0 +1,397 @@
+- READER should be funcall-able, so it should either be:
+  - A symbol for a function with one argument (an instance of the
+    class, as JSCValue).
+  - A lambda with one argument.
```
Nit: "of one argument" would be more precise I think.

Ambrevar commented on webkit2/jsc-utils.lisp:
```diff
@@ -0,0 +1,397 @@
+    class, as JSCValue).
+  - A lambda with one argument.
+- WRITER follows the same convention except that it should have a
+  second argument -- the value to set (the value is lispy)."
```
Maybe "the Lispy value to set".

Ambrevar commented on webkit2/jsc-utils.lisp:
```diff
@@ -0,0 +1,397 @@
+     class (or name (null-pointer)) callback
+     (null-pointer) (null-pointer) ;; TODO: Use g-notify-destroy-free?
+     jsc-value-type n-args
+     (if (not (zerop n-args))
```
You can remove the `not` and swap the two branches.

Ambrevar commented on webkit2/jsc-utils.lisp:
```diff
@@ -0,0 +1,397 @@
+(defmacro make-jsc-method (name ((class-var class) &rest args)
+                           &body body)
+  "Define a method NAME over CLASS.
+CLASS-VAR is bound to JSCValue of class CLASS when BODY is ran."
```
Typo: "run"

Ambrevar commented on webkit2/jsc-utils.lisp:
```diff
@@ -0,0 +1,397 @@
+CLASS-VAR is bound to JSCValue of class CLASS when BODY is ran."
+  (let ((method-name (typecase name
+                       (symbol (cffi:translate-camelcase-name name :upper-initial-p nil))
+                       (string name)))
```
Maybe prefer `etypecase` in case the use tries to bind NAME to something else?

Ambrevar commented on webkit2/tests/tests.lisp:
```diff
@@ -0,0 +1,418 @@
+
+(def-test everything-array-vector (:suite js-tests)
+  (setf webkit:*js-array-type* :vector
+        webkit:*js-object-type* :alist)
```
These `unwind-protect` are the unfortunate consequence that we cannot bind the dynamic variables properly, right?

What about making an (unexported) helper macro for now?  Maybe just for testing.

Ambrevar commented:
If you keep pushing changes, I'll keep sending my feedback :)

Looking great, have a look at my feedback, then we can merge.

Ambrevar commented:
I have an error when running the `zero-fn` test:

```lisp
The value
  NIL
is not of type
  CL-WEBKIT2:WEBKIT-WEB-VIEW
when binding CL-WEBKIT2::WEB-VIEW
   [Condition of type TYPE-ERROR]

Restarts:
 0: [ABORT] abort thread (#<THREAD "cl-cffi-gtk main thread" RUNNING {100491BC83}>)

Backtrace:
 0: (CL-WEBKIT2:WEBKIT-WEB-VIEW-EVALUATE-JAVASCRIPT NIL "null" #<FUNCTION (LAMBDA (CL-WEBKIT2::RESULT CL-WEBKIT2:JSC-VALUE) :IN CL-WEBKIT2:GET-JSC-CONTEXT) {53AB3AAB}> #<unavailable argument> #<unavailabl..
 1: ((LAMBDA NIL :IN CL-WEBKIT2:GET-JSC-CONTEXT))
```

Ambrevar commented:
About the docstring of `webkit-web-view-evaluate-javascript':

```
  "Evaluate JAVASCRIPT in WEB-VIEW (and WORLD, if present) calling CALL-BACK upon completion.
CALL-BACK is called over the result of Lisp transformation of result and original result JSCValue.
ERROR-CALL-BACK is called with the signaled condition."
```

- What is WORLD?
- Second sentence is unclear.  Maybe
  "CALL-BACK is called over two arguments:
  - The Lisp transformation of the result.
  - The untransformed result (a JSCValue)."

Ambrevar commented:
cl-webkit has a Git tag "2.4", that's why.

aartaka commented:
All addressed. The heisenbug with non-existent view was (again!) due to multithreading -- in some cases tests were ran earlier than initialization was done. cc7187b55669921e65cfd72cc701fe4b3eae2e44 fixes that.

aartaka commented:
I've not added a helper macro for js transformation parameters, though -- feels like an overkill.

Ambrevar commented:
> Won't work. Both `(funcall (lambda (a) a) 1 2)` and `(apply (lambda (a) a) 1 2 nil)` signal errors. We use strict one-argument callbacks like `(lambda (a) a)` everywhere, thus this code will become erroneous even with optional arguments. Thus, it's inevitably backwards incompatible.

I'm confused.  In you example, shouldn't you

(funcall (lambda (a &optional b) a) 1 2)

instead?

aartaka commented:
> Won't work. Both `(funcall (lambda (a) a) 1 2)` and `(apply (lambda (a) a) 1 2 nil)` signal errors. We use strict one-argument callbacks like `(lambda (a) a)` everywhere, thus this code will become erroneous even with optional arguments. Thus, it's inevitably backwards incompatible.
> I'm confused. In you example, shouldn't you (funcall (lambda (a &optional b) a) 1 2) instead?

What I mean is that all the code relying on `cl-webkit` already uses a _strictly one argument lambda_ ([Nyxt's renderer-gtk.lisp](https://github.com/atlas-engineer/nyxt/blob/d33257d4960953070e309ca1880168d17b8c0e14/source/renderer-gtk.lisp#L1030), for example). No matter how much optional arguments we require, the total number of arguments will always be more than 1 and existing code will error.

Ambrevar commented:
OK, got it! ;)

aartaka commented:
Small additions I could not resust a temptation to do:
- Stricted typing and better documentation of what variables do (85ac2c41731f05b5d608dff09e1d28c40b26bd0c).
- Number rounding configuration (48d031ae86d116544b1fa381390599909f059aa2).

The default behavior haven't changed an inch, tests pass, code is as clean and documented as I could possibly make it.

Ambrevar commented:
Feel free to merge :)

Ambrevar commented:
Don't forget to push a Git tag.

jmercouris commented:
It's great to see you two working so well together!

Ambrevar commented:
I've pushed the "3.0.0" tag because @aartaka pushed "3.0" instead which
does not match the version, thus making the Guix package definition a
bit awkward.

Shall we remove 3.0 then?

aartaka commented:
Ahhhhh, yes. Let's remove the 3.0 then!

Ambrevar commented:
Done!

aartaka commented:
Thanks, Pierre!
